src/main/java/me/zeroeightsix/kami/util/astaralg/Astarpathfinder.java

```diff
@@ -0,0 +1,650 @@
+                tunnel = Tunnels.ZTUNNEL;
+                break;
+            case 5:
+                //+x-z
+                target = new BlockPos(getPlayerBlockpos().getX() + 10, startplayerpos.y,
+                        startplayerpos.z - Math.abs(startplayerpos.x - (getPlayerBlockpos().getX() + 10)));
+                tunnel = Tunnels.PNNPTUNNEL;
+                break;
+            case 6:
+                //+x
+                target = new BlockPos(getPlayerBlockpos().getX() + 10, startplayerpos.y,
+                        startplayerpos.z);
+                tunnel = Tunnels.XTUNNEL;
+                break;
+            case 7:
+                //+x+z
+                target = new BlockPos(getPlayerBlockpos().getX() + 10, startplayerpos.y, startplayerpos.z + Math.abs(startplayerpos.x - (getPlayerBlockpos().getX() + 10)));
+                tunnel = Tunnels.PPNNTUNNEL;
+                break;
+            default:
+                target = getPlayerBlockpos();
+                tunnel = Tunnels.XTUNNEL;
+                break;
+        }
+        Iterable<BlockPos> blocks = BlockPos.getAllInBox(new BlockPos(target.x - 20, Math.max(3, target.y - 10), target.z - 20), new BlockPos(target.x + 20, Math.min(roof, target.y + 10), target.z + 20));
+        List<BlockPos> blocklist = new ArrayList<>();
+        blocks.iterator().forEachRemaining(blocklist::add);
+        blocklist.sort(Comparator.comparingDouble(o -> Math.pow(target.x - o.x, 2) + Math.pow((target.y - o.y) * 3, 2) + Math.pow(target.z - o.z, 2)));
+        for (BlockPos b : blocklist) {
+            if (checkTunnelAt(b, tunnel)) {
+                return b;
+            }
+        }
+        MessageSendHelper.sendChatMessage("No tunnel Found");
+        return target;
+
+    }
+
+    private static boolean checkTunnelAt(BlockPos center, List<Vec3i> tunnel) {
+        boolean blocked = true;
+        if (Blocks.BEDROCK.equals(mc.world.getBlockState(center.add(0, 2, 0)).getBlock())
+                || Blocks.LAVA.equals(mc.world.getBlockState(center.add(0, 3, 0)).getBlock())
+                || Blocks.FLOWING_LAVA.equals(mc.world.getBlockState(center.add(0, 3, 0)).getBlock())
+                || Blocks.LAVA.equals(mc.world.getBlockState(center.add(0, -1, 0)).getBlock())
+                || Blocks.FLOWING_LAVA.equals(mc.world.getBlockState(center.add(0, -1, 0)).getBlock())
+                || MyAstarNode.noclipblocks.contains(mc.world.getBlockState(center.add(0, -1, 0)).getBlock())
+        ) {
+            return false;
+        }
+        for (Vec3i v : tunnel) {
+            if (!MyAstarNode.noclipblocks.contains(mc.world.getBlockState(center.add(v)).getBlock())) {
+                blocked = false;
+                break;
+            }
+        }
+        return blocked;
+    }
+
+    private static BlockPos getPlayerBlockpos() {
+        return new BlockPos((int) Math.floor(mc.player.posX), (int) Math.floor(mc.player.posY), (int) Math.floor(mc.player.posZ));
+    }
+}
```

Comment:
I'll probably rewrite this in Kotlin once everything else is done

Reply:
could def profit form the nice when syntax. I just didnt feel secure enough in kotlin for now